go/oasis-test-runner/scenario/e2e/runtime/runtime_upgrade.go

```diff
@@ -106,9 +109,80 @@ func (sc *runtimeUpgradeImpl) Run(ctx context.Context, childEnv *env.Env) error
 	// Run client again.
 	sc.Logger.Info("starting a second client to check if runtime works")
 	sc.Scenario.TestClient = NewTestClient().WithSeed("seed2").WithScenario(InsertRemoveEncWithSecretsScenarioV2)
+
+	// Ensure that after upgrade, every compute worker had its old exploded
+	// bundle removed from its bundles dir.
+	for _, worker := range sc.Net.ComputeWorkers() {
+		if err := sc.verifyBundleDir(ctx, worker); err != nil {
+			sc.Logger.Error("compute worker bundle dir clean-up error",
+				"worker", worker.Name,
+				"err", err,
+			)
+			return err
+		}
+	}
+
 	return sc.RunTestClientAndCheckLogs(ctx, childEnv)
 }
 
+func (sc *runtimeUpgradeImpl) verifyBundleDir(ctx context.Context, worker *oasis.Compute) error {
+	sc.Logger.Info("ensuring cached exploded bundle for version 0.0.0 was removed",
+		"worker", worker.Name,
+	)
+
+	// There should be only one exploded bundle dir.
+	bundlesDir := bundle.ExplodedPath(worker.DataDir())
+	entries, err := os.ReadDir(bundlesDir)
+	if err != nil {
+		return err
+	}
+	// if n := len(entries); n != 1 {
+	// 	return fmt.Errorf("unexpected number of dir entries: expected 1, got %d", n)
+	// }
+	entry := entries[0]
+	if !entry.IsDir() {
+		return fmt.Errorf("%s is not a dir", entry.Name())
+	}
+
+	// Ensure exploded cached bundle is for the latest version (0.1.0).
+	// We do so by getting the manifest hash of the active deployment,
+	// and ensuring there is a dir with corresponding hash.
+
+	// Fetch registry descriptor.
+	rt, err := sc.Net.Controller().Registry.GetRuntime(ctx, &registry.GetRuntimeQuery{
+		Height: consensus.HeightLatest,
+		ID:     sc.Net.Runtimes()[sc.upgradedRuntimeIndex].ID(),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get runtime descriptor: %w", err)
+	}
+	// Fetch current epoch.
+	epoch, err := sc.Net.Controller().Beacon.GetEpoch(ctx, consensus.HeightLatest)
+	if err != nil {
+		return fmt.Errorf("failed to get current epoch: %w", err)
+	}
+	// Get active deployment manifest hash.
+	active := rt.ActiveDeployment(epoch)
+	if active == nil {
+		return fmt.Errorf("missing active descriptor for epoch %v", epoch)
+	}
+	manifestHash := active.BundleChecksum
+
+	// Ensure single entry name (exploded bundle dir) equals to fetched manifest hash.
+	var want hash.Hash
+	if err = want.UnmarshalBinary(manifestHash); err != nil {
+		return fmt.Errorf("failed to unmarshal active deployment manifest hash")
+	}
+	var got hash.Hash
+	if err := got.UnmarshalHex(entry.Name()); err != nil {
+		return fmt.Errorf("failed to unmarshal dir name to hash")
+	}
+	if !want.Equal(&got) {
+		return fmt.Errorf("unexpected exploded bundle hash: want %v, got %v", want, got)
```

Comment:
This comment could be improved, to say that the folder name is not correct or that folder content failed to verify, as this is what we are testing.